saleor/webhook/observability_reporter/obfuscation.py

```diff
@@ -0,0 +1,193 @@
+]
+MASK = "***"
+
+
+def hide_sensitive_headers(
+    headers: Dict[str, str], sensitive_headers: Tuple[str, ...] = SENSITIVE_HEADERS
+) -> Dict[str, str]:
+    return {
+        key: val if key.upper().replace("-", "_") not in sensitive_headers else MASK
+        for key, val in headers.items()
+    }
+
+
+class SensitiveFieldError(GraphQLError):
+    pass
+
+
+class ContainSensitiveField(ValidationRule):
+    def __init__(
+        self, sensitive_fields: SensitiveFieldsMap
+    ):  # pylint: disable=super-init-not-called
+        self.sensitive_fields = sensitive_fields
+
+    def __call__(self, context: ValidationContext):
+        self.context = context
+        return self
+
+    def is_sensitive_field(self, node: Field, parent_type: str):
+        if fields := self.sensitive_fields.get(parent_type):
+            node_name = node.name.value
+            if node_name in fields:
+                raise SensitiveFieldError(
+                    "The query contains sensitive field "
+                    f"{node_name} of type {parent_type}."
+                )
+
+    def contain_sensitive_field(self, node: GraphQLNode, type_def) -> bool:
+        if isinstance(node, FragmentSpread) or not node.selection_set:
+            return False
+        fields: Dict[str, GraphQLField] = {}
+        if isinstance(type_def, (GraphQLObjectType, GraphQLInterfaceType)):
+            fields = type_def.fields
+        for child_node in node.selection_set.selections:
+            if isinstance(child_node, Field):
+                field = fields.get(child_node.name.value)
+                if not field:
+                    continue
+                field_type = get_named_type(field.type)
+                if type_def and type_def.name:
+                    self.is_sensitive_field(child_node, type_def.name)
+                self.contain_sensitive_field(child_node, field_type)
+            if isinstance(child_node, FragmentSpread):
+                fragment = self.context.get_fragment(child_node.name.value)
+                if fragment:
+                    fragment_type = self.context.get_schema().get_type(
+                        fragment.type_condition.name.value
+                    )
+                    self.contain_sensitive_field(fragment, fragment_type)
+            if isinstance(child_node, InlineFragment):
+                inline_fragment_type = type_def
+                if child_node.type_condition and child_node.type_condition.name:
+                    inline_fragment_type = self.context.get_schema().get_type(
+                        child_node.type_condition.name.value
+                    )
+                self.contain_sensitive_field(child_node, inline_fragment_type)
+        return False
+
+    def enter_operation_definition(
+        self, node, key, parent, path, ancestors
+    ):  # pylint: disable=unused-argument
+        validate_sensitive_fields_map(self.sensitive_fields, self.context.get_schema())
+        if node.operation == "query":
```

Comment:
Looks like you call `contain_sensitive_field` for all `if` but with different condition. Maybe we could use `if` to gather specific params (like `.get_query_type()`) and call `contain_sensitive_field` only once after processing the `if`s?